system/jhtree/jhtree.cpp

```diff
@@ -2508,150 +2518,110 @@ CJHTreeNode *CNodeCache::getNode(INodeLoader *keyIndex, unsigned iD, offset_t po
 
     // No benefit in caching the following, especially since they will evict useful pages
     if ((type == NodeMeta) || (type == NodeBloom))
-        return keyIndex->loadNode(nullptr, nullptr, pos);
+        return keyIndex->loadNode(nullptr, pos);
 
     //NOTE: TLK leaf nodes are currently cached along with branches, not with leaves.  It might be better if this was a separate cache.
     CacheType cacheType = isTLK ? CacheBranch : (CacheType)type;
 
     // check cacheEnabled[cacheType] avoid the critical section (and testing the flag within the critical section)
     if (unlikely(!cacheEnabled[cacheType]))
-        return keyIndex->loadNode(nullptr, nullptr, pos);
+        return keyIndex->loadNode(nullptr, pos);
 
-    //Legacy cache access:
+    //Previously, this was implemented as:
     //  Lock, unlock.  Load the page.  Lock, check if it has been added, otherwise add.
-    //New code:
+    //Now, it is coded as:
     //  Lock, add if missing, unlock.  Lock a page-dependent-cr load() release lock.
     //There will be the same number of critical section locks, but loading a page will contend on a different lock - so it should reduce contention.
     //There will be a limit on the number of nodes concurrently being loaded from memory with the new code, where it was unlimited before, but
     //nodes will only be loaded once.
     CKeyIdAndPos key(iD, pos);
     CriticalSection & cacheLock = lock[cacheType];
-    if (legacyMode)
+    Owned<CNodeCacheEntry> ownedCacheEntry; // ensure node gets cleaned up if it fails to load
```

Comment:
I would also have
```
Owned<const CJHTreeNode *> ownedMatch;
...
if (cacheEntry->isReady())
   ownedMatch.setown(cacheEntry->queryNode())
else
  ownedCacheEntry.set(cacheEntry);

```
That will save an atomic link and release on the cache entry for the common case where it is already in the cache.

Reply:
Leave to a later PR